src/checkers/inference/solver/backend/logiqlbackend/LogiQLPredicateGenerator.java

```diff
@@ -0,0 +1,190 @@
+        //predicates for making output in order.
+        basicEncoding.append("variableOrder(v) -> int(v).\n");
+        basicEncoding.append("variableOrder(i) <- variable(v), hasvariableName(v:i).\n");
+        basicEncoding.append("orderVariable[o] = v -> int(o), int(v).\n");
+        basicEncoding.append("orderVariable[o] = v <- seq<<o=v>> variableOrder(v).\n");
+        basicEncoding.append("orderedAnnotationOf[v] = a -> int(v), string(a).\n");
+        basicEncoding.append("orderedAnnotationOf[v] = a <- variable(q), hasvariableName(q:v), AnnotationOf[q]=a, orderVariable[_]=v.\n");
+        //predicates for constraints.
+        //equality constraint
+        basicEncoding.append("equalityConstraint(v1,v2) -> variable(v1), variable(v2).\n");
+        basicEncoding.append("equalityConstraintContainsConstant(v1,v2) -> constant(v1), variable(v2).\n");
+        //inequality constraint
+        basicEncoding.append("inequalityConstraint(v1,v2) -> variable(v1), variable(v2).\n");
+        basicEncoding.append("inequalityConstraintContainsConstant(v1,v2) -> constant(v1), variable(v2).\n");
+        //subtype constraint
+        basicEncoding.append("subtypeConstraint(v1,v2) -> variable(v1), variable(v2).\n");
+        basicEncoding.append("subtypeConstraintLeftConstant(v1,v2) -> constant(v1), variable(v2).\n");
+        basicEncoding.append("subtypeConstraintRightConstant(v1,v2) -> variable(v1), constant(v2).\n");
+        //comparable constraint
+        basicEncoding.append("comparableConstraint(v1,v2) -> variable(v1), variable(v2).\n");
+        basicEncoding.append("comparableConstraintContainsConstant(v1,v2) -> constant(v1), variable(v2).\n");
+        // each type
+        for (AnnotationMirror annoMirror : lattice.allTypes) {
+            basicEncoding.append("is" + NameUtils.getSimpleName(annoMirror)
+                    + "[v] = i -> variable(v), boolean(i).\n");
+        }
+        for (AnnotationMirror annoMirror : lattice.allTypes) {
+            String simpleName = NameUtils.getSimpleName(annoMirror);
+            basicEncoding.append("AnnotationOf[v] = \"" + simpleName + "\" <- " + "is" + simpleName + "[v] = true.\n");
+        }
+        for (AnnotationMirror rightAnnoMirror : lattice.allTypes) {
+            for (AnnotationMirror leftAnnoMirror : lattice.allTypes) {
+                String leftAnnoName = NameUtils.getSimpleName(leftAnnoMirror);
+                String rightAnnoName = NameUtils.getSimpleName(rightAnnoMirror);
+                if (!leftAnnoName.equals(rightAnnoName)) {
+                    basicEncoding.append("is" + leftAnnoName + "[v] = false <- is" + rightAnnoName + "[v] = true.\n");
+                }
+                
+            }
+        }
+        return basicEncoding.toString();
+    }
+    
+    /**
+     * write all encoding generated by this class to file LogiqlEncoding.logic.
+     *
+     *
+     */
+    private void writeFile(String output) {
+        String[] lines = output.split("\r\n|\r|\n");
```

Comment:
The way of using StringBuilder and then split the string it builds seem rather strange.

Also it is tricky for Serializer adding `\n` on each constraint just for here to split them up.

Refactor this.